boa/src/builtins/set/ordered_set.rs

```diff
@@ -0,0 +1,130 @@
+use crate::gc::{custom_trace, Finalize, Trace};
+use indexmap::{
+    set::{IntoIter, Iter},
+    IndexSet,
+};
+use std::{
+    collections::hash_map::RandomState,
+    fmt::Debug,
+    hash::{BuildHasher, Hash},
+};
+
+/// A newtype wrapping indexmap::IndexSet
+#[derive(Clone)]
+pub struct OrderedSet<V, S = RandomState>(IndexSet<V, S>)
+where
+    V: Hash + Eq;
+
+impl<V: Eq + Hash + Trace, S: BuildHasher> Finalize for OrderedSet<V, S> {}
+unsafe impl<V: Eq + Hash + Trace, S: BuildHasher> Trace for OrderedSet<V, S> {
+    custom_trace!(this, {
+        for v in this.0.iter() {
+            mark(v);
+        }
+    });
+}
```

Comment:
Since this is implementing a custom trace, I would add a comment on why this is needed and how it works.

Reply:
Hmmm maybe @HalidOdat knows about this more than me.

Reply:
> Hmmm maybe @HalidOdat knows about this more than me.

It is needed because `IndexSet` is does not implement `Trace`, that's why we wrap it in a struct that does implement it (`OrderedSet`). The custom trace iterates through all the values in the set and calls trace on them (the `mark(v)` part).